### PR TITLE
sourmash: new package @4.8.2

### PR DESCRIPTION
**Pull request by A-N-Other**
- Adding `sourmash@4.8.2`. This is a CLI package rather than a library, so I'm dropping the `py-` prefix. This package builds with the `maturin` backend, which has taken a while to get to work here - some manual fixes were required to properly deploy the built library and to correct an auto-generated CFFI file.
- `py-bitstring` has moved to shipping tarballs instead of zips and has a new dependency as of `@4.1:`.
- `py-maturin` needed updating.
- `py-bitarray` is required at a specific version for `py-bitstring`.
- The `py-screed` dependency is a new package.

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-bitstring/package.py [hunk 1]
```diff
@@ -10,9 +10,15 @@ class PyBitstring(PythonPackage):
...
     version("3.1.5", sha256="c163a86fcef377c314690051885d86b47419e3e1770990c212e16723c1c08faa")
 
-    # pip silently replaces distutils with setuptools
     depends_on("py-setuptools", type="build")
+    depends_on("py-bitarray@2.7.4", when="@4.1:")
+
+    def url_for_version(self, version):
+        if version < Version("3.1.6"):
+            url = "https://pypi.org/packages/source/s/bitstring/bitstring-{0}.zip"
+            return url.format(version)
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-bitstring/package.py]: This needs to work for all versions. I would use something like:
```suggestion
        url = "https://pypi.org/packages/source/s/bitstring/bitstring-{}.{}"
        if version < Version("3.1.6"):
            suffix = "zip"
        else:
            suffix = "tar.gz"
        return url.format(version, suffix)
```

File: var/spack/repos/builtin/packages/py-bitstring/package.py [hunk 2]
```diff
@@ -10,9 +10,15 @@ class PyBitstring(PythonPackage):
...
 
     homepage = "http://pythonhosted.org/bitstring"
-    pypi = "bitstring/bitstring-3.1.5.zip"
+    pypi = "bitstring/bitstring-3.1.6.tar.gz"
 
+    version("4.0.2", sha256="a391db8828ac4485dd5ce72c80b27ebac3e7b989631359959e310cd9729723b2")
     version("3.1.5", sha256="c163a86fcef377c314690051885d86b47419e3e1770990c212e16723c1c08faa")
 
-    # pip silently replaces distutils with setuptools
     depends_on("py-setuptools", type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-bitstring/package.py]: setuptools 61+ is required now

File: var/spack/repos/builtin/packages/py-bitstring/package.py [hunk 3]
```diff
@@ -10,9 +10,15 @@ class PyBitstring(PythonPackage):
...
+    depends_on("py-bitarray@2.7.4", when="@4.1:")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-bitstring/package.py:20]: Where did you find this dep?

File: var/spack/repos/builtin/packages/py-screed/package.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+class PyScreed(PythonPackage):
+    """screed: a Python library for loading FASTA and FASTQ sequences."""
+
+    homepage = "https://screed.readthedocs.io/"
+    pypi = "screed/screed-1.1.2.tar.gz"
+
+    version("1.1.2", sha256="734ffa7a8a645286496d895b736f91d6b2988956e2fd42358123d93ec8519b6a")
+
+    depends_on("py-setuptools", type="build")
+    depends_on("py-setuptools-scm", type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-screed/package.py]: All of these build deps and versions need to be added:
```toml
requires = [
    "setuptools >= 48",
    "setuptools_scm[toml] >= 4, <6",
    "setuptools_scm_git_archive",
    "wheel >= 0.29.0",
]
```

File: var/spack/repos/builtin/packages/sourmash/package.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    k-mer sketches."""
+
+    homepage = "https://sourmash.bio/"
+    pypi = "sourmash/sourmash-4.8.2.tar.gz"
+
+    version("4.8.2", sha256="e0df78032e53ed88977445933ba3481dd10c7d3bd26d019511a6a4e6d7518475")
+
+    depends_on("python@3.8:")
+    # build-only
+    depends_on("py-maturin@1:", type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/sourmash/package.py]:
```suggestion
    depends_on("py-maturin@0.14.13:0.14", type="build")
```

File: var/spack/repos/builtin/packages/sourmash/package.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    depends_on("rust", type="build")
+    # general
+    depends_on("py-screed@1.1.2:")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/sourmash/package.py]:
```suggestion
    depends_on("py-screed@1.1.2:1")
```

File: var/spack/repos/builtin/packages/sourmash/package.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    depends_on("py-cffi@1.14.0:")
+    depends_on("py-numpy")
+    depends_on("py-matplotlib")
+    depends_on("py-scipy")
+    depends_on("py-deprecation@2.0.6:")
+    depends_on("py-cachetools@4:")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/sourmash/package.py]:
```suggestion
    depends_on("py-cachetools@4:5")
```

File: var/spack/repos/builtin/packages/sourmash/package.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    depends_on("py-bitstring@3.1.9:")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/sourmash/package.py]:
```suggestion
    depends_on("py-bitstring@3.1.9:4")
```

File: var/spack/repos/builtin/packages/sourmash/package.py [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    depends_on("py-importlib_metadata@3.6:", when="^python@:3.9")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/sourmash/package.py]: These all need to be `type=("build", "run")`

File: var/spack/repos/builtin/packages/sourmash/package.py [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+        cargo = Executable("cargo")
+        cargo("build", "--release")
+        # install python package
+        args = std_pip_args + ["--prefix=" + prefix, "."]
+        pip(*args)
+        # move sourmash.so into expected place
+        site_packages = join_path(
+            prefix.lib,
+            "python{0}".format(spec["python"].version.up_to(2)),
+            "site-packages",
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/sourmash/package.py]: This is equivalent to `python_platlib`, I would use that instead. Sometimes `lib64` is needed instead

File: var/spack/repos/builtin/packages/sourmash/package.py [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+class Sourmash(PythonPackage):
+    """Sourmash: Quickly search, compare, and analyze genomic and metagenomic data sets with
+    k-mer sketches."""
+
+    homepage = "https://sourmash.bio/"
+    pypi = "sourmash/sourmash-4.8.2.tar.gz"
+
+    version("4.8.2", sha256="e0df78032e53ed88977445933ba3481dd10c7d3bd26d019511a6a4e6d7518475")
+
+    depends_on("python@3.8:")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/sourmash/package.py]:
```suggestion
    depends_on("python@3.8:", type=("build", "run"))
```
Unless you also need to link to it

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-bitstring/package.py (hunk 2 above)
Comment (A-N-Other) [reply var/spack/repos/builtin/packages/py-bitstring/package.py]: Done

File: var/spack/repos/builtin/packages/py-bitstring/package.py (hunk 3 above)
Comment (A-N-Other) [reply var/spack/repos/builtin/packages/py-bitstring/package.py:20]: Sorry, should have said in the PR comment. They're switching over to use `bitarray` as a backend as of `@4.1:`: https://github.com/scott-griffiths/bitstring#41-beta-now-available. That's not the version I'm adding as it's still in beta (`when="@4.1:"` > `@4.0.2`), but I was just including it now for completeness. The `main` branch has this dep in its `pyproject.toml`.

File: var/spack/repos/builtin/packages/py-screed/package.py (hunk 4 above)
Comment (A-N-Other) [reply var/spack/repos/builtin/packages/py-screed/package.py]: Done. Missed a `git add` somewhere along the line, I think :/

File: var/spack/repos/builtin/packages/sourmash/package.py (hunk 5 above)
Comment (A-N-Other) [reply var/spack/repos/builtin/packages/sourmash/package.py]: Think I must have been looking at `main`, as they've switched to `@1:` there. Good spot.

File: var/spack/repos/builtin/packages/sourmash/package.py (hunk 6 above)
Comment (A-N-Other) [reply var/spack/repos/builtin/packages/sourmash/package.py]: Had left it as `2` doesn't exist yet. Committed

File: var/spack/repos/builtin/packages/sourmash/package.py (hunk 7 above)
Comment (A-N-Other) [reply var/spack/repos/builtin/packages/sourmash/package.py]: Had left it as `6` doesn't exist yet. Committed

File: var/spack/repos/builtin/packages/sourmash/package.py (hunk 8 above)
Comment (A-N-Other) [reply var/spack/repos/builtin/packages/sourmash/package.py]: As with the others, they've only just released `@4`, so `5` is ages away. Committed.

File: var/spack/repos/builtin/packages/sourmash/package.py (hunk 9 above)
Comment (A-N-Other) [reply var/spack/repos/builtin/packages/sourmash/package.py]: They do indeed. Done.

File: var/spack/repos/builtin/packages/sourmash/package.py (hunk 10 above)
Comment (A-N-Other) [reply var/spack/repos/builtin/packages/sourmash/package.py]: Thanks, didn't know about that one. Switched that over and done a test install - all good.

File: var/spack/repos/builtin/packages/py-bitstring/package.py (hunk 3 above)
Comment (A-N-Other) [reply var/spack/repos/builtin/packages/py-bitstring/package.py:20]: Can remove it for now if you prefer...

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-bitstring/package.py:20]: Gotcha, didn't realize it was for future versions

File: var/spack/repos/builtin/packages/py-bitarray/package.py [hunk 12]
```diff
@@ -11,6 +11,8 @@ class PyBitarray(PythonPackage):
 
     pypi = "bitarray/bitarray-0.8.1.tar.gz"
 
+    version("2.7.6", sha256="3807f9323c308bc3f9b58cbe5d04dc28f34ac32d852999334da96b42f64b5356")
+    version("2.7.4", sha256="143d4f65e1f45a533e13521be1dc557a782317ecf76520eabd5a903b26ecb187")
     version("2.6.0", sha256="56d3f16dd807b1c56732a244ce071c135ee973d3edc9929418c1b24c5439a0fd")
     version("0.8.1", sha256="7da501356e48a83c61f479393681c1bc4b94e5a34ace7e08cb29e7dd9290ab18")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-bitarray/package.py:18]: You can remove `depends_on("python")`, it's already added by the base class

File: var/spack/repos/builtin/packages/py-screed/package.py [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+class PyScreed(PythonPackage):
+    """screed: a Python library for loading FASTA and FASTQ sequences."""
+
+    homepage = "https://screed.readthedocs.io/"
+    pypi = "screed/screed-1.1.2.tar.gz"
+
+    version("1.1.2", sha256="734ffa7a8a645286496d895b736f91d6b2988956e2fd42358123d93ec8519b6a")
+
+    depends_on("py-setuptools@48:", type="build")
+    depends_on("py-setuptools-scm@4:5.0.2", type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-screed/package.py]:
```suggestion
    depends_on("py-setuptools-scm@4:5", type="build")
```

File: var/spack/repos/builtin/packages/py-bitarray/package.py (hunk 12 above)
Comment (A-N-Other) [reply var/spack/repos/builtin/packages/py-bitarray/package.py:18]: Done

**Review by adamjstewart — DISMISSED** [review]

**Comment by alalazo** [comment]
Apparently there are conflicts to be solved (though GitHub UI won't say where they are)

**Comment by A-N-Other** [comment]
Should have that merge conflict fixed now @alalazo @adamjstewart
Thanks!

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**